src/main/java/org/radarcns/management/web/rest/OAuthClientsResource.java

```diff
@@ -80,20 +93,110 @@
+        return ResponseEntity.ok().body(clientDetailsMapper
+                .clientDetailsToClientDetailsDTO(getOAuthClient(id)));
+    }
+
+    /**
+     * PUT /api/oauthclients
+     *
+     * Update an existing OAuth client.
+     *
+     * @param clientDetailsDTO The client details to update
+     * @return The updated OAuth client.
+     */
+    @PutMapping("/oauthclients")
+    @Timed
+    public ResponseEntity<ClientDetailsDTO> updateOAuthClient(@RequestBody ClientDetailsDTO
+            clientDetailsDTO) {
+        checkPermission(getJWT(servletRequest), OAUTHCLIENTS_UPDATE);
+        checkProtected(getOAuthClient(clientDetailsDTO.getClientId()));
+        ClientDetails details = clientDetailsMapper
+                .clientDetailsDTOToClientDetails(clientDetailsDTO);
+        clientDetailsService.updateClientDetails(details);
+        ClientDetails updated = getOAuthClient(clientDetailsDTO.getClientId());
+        // updateClientDetails does not update secret, so check for it seperately
+        if (Objects.nonNull(clientDetailsDTO.getClientSecret()) && !clientDetailsDTO
+                .getClientSecret().equals(updated.getClientSecret())) {
+            clientDetailsService.updateClientSecret(clientDetailsDTO.getClientId(),
+                    clientDetailsDTO.getClientSecret());
+        }
+        updated = getOAuthClient(clientDetailsDTO.getClientId());
+        return ResponseEntity.ok()
+                .headers(HeaderUtil.createEntityUpdateAlert(ENTITY_NAME,
+                        clientDetailsDTO.getClientId()))
+                .body(clientDetailsMapper.clientDetailsToClientDetailsDTO(updated));
+    }
+
+    /**
+     * DELETE /api/oauthclients/:id
+     *
+     * Delete the OAuth client with the specified client id.
+     *
+     * @param id The id of the client to delete
+     * @return a ResponseEntity indicating success or failure
+     */
+    @DeleteMapping("/oauthclients/{id}")
+    @Timed
+    public ResponseEntity<Void> deleteOAuthClient(@PathVariable String id) {
+        checkPermission(getJWT(servletRequest), OAUTHCLIENTS_DELETE);
+        checkProtected(getOAuthClient(id));
+        clientDetailsService.removeClientDetails(id);
+        return ResponseEntity.ok().headers(HeaderUtil.createEntityDeletionAlert(ENTITY_NAME, id))
+                .build();
+    }
+
+    /**
+     * POST /api/oauthclients
+     *
+     * Register a new oauth client
+     *
+     * @param clientDetailsDTO The oauth client to be registered
+     * @return a response indicating success or failure
+     * @throws URISyntaxException if there was a problem formatting the URI to the new entity
+     */
+    @PostMapping("/oauthclients")
+    @Timed
+    public ResponseEntity<ClientDetailsDTO> createOAuthClient(@RequestBody ClientDetailsDTO
+            clientDetailsDTO) throws URISyntaxException {
+        checkPermission(getJWT(servletRequest), OAUTHCLIENTS_CREATE);
+        ClientDetails details = clientDetailsMapper
```

Comment:
Should check whether  `client-id` is null and  whether there a client already exist with current `client-id`.  Similar to 
```
        if (sourceDTO.getId() != null) {
            return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert(ENTITY_NAME, "idexists", "A new source cannot already have an ID")).body(null);
        } else if (sourceDTO.getSourceId() != null) {
            return ResponseEntity.badRequest()
                .headers(HeaderUtil.createFailureAlert(ENTITY_NAME, "sourceIdExists", "A new source cannot already have a Source ID"))
                .body(null);
        } else if (sourceRepository.findOneBySourceName(sourceDTO.getSourceName()).isPresent()) {
            return ResponseEntity.badRequest()
                .headers(HeaderUtil.createFailureAlert(ENTITY_NAME, "sourceNameExists", "Source name already in use"))
                .body(null);
        }
```